File: backend/cola circular.py
class ColaCircular:
    def __init__(self, capacidad=10):
        self.capacidad = capacidad
        self.datos = [None] * capacidad
        self.frente = 0
        self.final = 0
        self.tamano = 0

    def esta_llena(self):
        return self.tamano == self.capacidad

    def esta_vacia(self):
        return self.tamano == 0

    def enqueue(self, cliente):
        if self.esta_llena():
            return False
        self.datos[self.final] = cliente
        self.final = (self.final + 1) % self.capacidad
        self.tamano += 1
        return True

    def dequeue(self):
        if self.esta_vacia():
            return None
        cliente = self.datos[self.frente]
        self.datos[self.frente] = None
        self.frente = (self.frente + 1) % self.capacidad
        self.tamano -= 1
        return cliente

    def ver_cola(self):
        if self.esta_vacia():
            return []
        resultado = []
        i = self.frente
        for _ in range(self.tamano):
            resultado.append(self.datos[i])
            i = (i + 1) % self.capacidad
        return resultado

File: backend/colacircular.js
class ColaCircular {
  constructor(capacidad = 10) {
    this.items = new Array(capacidad);
    this.capacidad = capacidad;
    this.frente = 0;
    this.final = -1;
    this.tamano = 0;
  }

  estaVacia() {
    return this.tamano === 0;
  }

  estaLlena() {
    return this.tamano === this.capacidad;
  }

  enqueue(cliente) {
    if (this.estaLlena()) {
      return {
        exito: false,
        mensaje: "La cola está llena"
      };
    }

    this.final = (this.final + 1) % this.capacidad;
    this.items[this.final] = cliente;
    this.tamano++;

    return {
      exito: true,
      mensaje: "Cliente agregado a la cola",
      cliente
    };
  }

  dequeue() {
    if (this.estaVacia()) {
      return {
        exito: false,
        mensaje: "La cola está vacía",
        cliente: null
      };
    }

    const cliente = this.items[this.frente];
    this.items[this.frente] = null;
    this.frente = (this.frente + 1) % this.capacidad;
    this.tamano--;

    return {
      exito: true,
      mensaje: "Cliente atendido",
      cliente
    };
  }

  verFrente() {
    if (this.estaVacia()) {
      return null;
    }

    return this.items[this.frente];
  }

  obtenerCola() {
    const cola = [];

    for (let i = 0; i < this.tamano; i++) {
      const posicion = (this.frente + i) % this.capacidad;
      cola.push(this.items[posicion]);
    }

    return cola;
  }

  obtenerTamano() {
    return this.tamano;
  }

  limpiar() {
    this.items = new Array(this.capacidad);
    this.frente = 0;
    this.final = -1;
    this.tamano = 0;
  }
}

module.exports = ColaCircular;